### PR TITLE
stm32h7 Add serial RX and TX DMA

### PR DESCRIPTION
**Pull request by davids5**
## Summary

This adds DMA support to the stm32h7 serial driver.

## Impact

Lowers load a 3Mbps

## Testing

px4_fmu-v6x, nucleo-h743zi

### CODE REVIEW

File: arch/arm/src/stm32h7/stm32_serial.c [hunk 1]
```diff
@@ -63,6 +65,410 @@
...
+#    if (defined(CONFIG_USART1_TXDMA) && defined(CONFIG_USART1_RS485)) || \
+        (defined(CONFIG_USART2_TXDMA) && defined(CONFIG_USART2_RS485)) || \
+        (defined(CONFIG_USART3_TXDMA) && defined(CONFIG_USART3_RS485)) || \
+        (defined(CONFIG_UART4_TXDMA) && defined(CONFIG_UART4_RS485)) || \
+        (defined(CONFIG_UART5_TXDMA) && defined(CONFIG_UART5_RS485)) || \
+        (defined(CONFIG_USART6_TXDMA) && defined(CONFIG_USART6_RS485)) || \
+        (defined(CONFIG_UART7_TXDMA) && defined(CONFIG_UART7_RS485)) || \
+        (defined(CONFIG_UART8_TXDMA) && defined(CONFIG_UART8_RS485))
+#      error "TXDMA and RS-485 cannot be enabled at the same time for the same U[S]ART"
+#    endif
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_serial.c:353]:
```suggestion
#if (defined(CONFIG_USART1_TXDMA) && defined(CONFIG_USART1_RS485)) || \
     (defined(CONFIG_USART2_TXDMA) && defined(CONFIG_USART2_RS485)) || \
     (defined(CONFIG_USART3_TXDMA) && defined(CONFIG_USART3_RS485)) || \
     (defined(CONFIG_UART4_TXDMA) && defined(CONFIG_UART4_RS485)) || \
     (defined(CONFIG_UART5_TXDMA) && defined(CONFIG_UART5_RS485)) || \
     (defined(CONFIG_USART6_TXDMA) && defined(CONFIG_USART6_RS485)) || \
     (defined(CONFIG_UART7_TXDMA) && defined(CONFIG_UART7_RS485)) || \
     (defined(CONFIG_UART8_TXDMA) && defined(CONFIG_UART8_RS485))
#  error "TXDMA and RS-485 cannot be enabled at the same time for the same U[S]ART"
#endif
```

File: arch/arm/src/stm32h7/stm32_serial.c [hunk 2]
```diff
@@ -2028,6 +2947,322 @@ static bool up_rxflowcontrol(struct uart_dev_s *dev,
...
+
+  uart_xmitchars_done(&priv->dev);
+}
+#endif
+
+/****************************************************************************
+ * Name: up_dma_txavailable
+ *
+ * Description:
+ *        Informs DMA that Tx data is available and is ready for transfer.
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_serial.c:3188]:
```suggestion
 *   Informs DMA that Tx data is available and is ready for transfer.
```

File: arch/arm/src/stm32h7/stm32_serial.c [hunk 3]
```diff
@@ -63,6 +65,410 @@
 
 #define STM32_NSERIAL (STM32H7_NUSART + STM32H7_NUART)
 
+/* DMA configuration */
+
+/* If DMA is enabled on any USART, then very that other pre-requisites
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_serial.c:70]:
```suggestion
/* If DMA is enabled on any USART, then verify that other pre-requisites
```

File: arch/arm/src/stm32h7/stm32_serial.c [hunk 4]
```diff
@@ -2324,10 +3619,10 @@ void arm_serialinit(void)
   int ret;
 #endif
 
+#ifdef CONFIG_PM
   /* Register to receive power management callbacks */
 
-#ifdef CONFIG_PM
-  ret = pm_register(&g_serialcb);
+  ret = pm_register(&g_serialpm.pm_cb);
```
Comment (raiden00pl) [inline arch/arm/src/stm32h7/stm32_serial.c:3625]: g_serialpm is not defined in this file. Probably `static  struct pm_callback_s g_serialcb` should be replaced with `static struct pm_config_s g_serialpm`

**Review by raiden00pl — CHANGES_REQUESTED** [review]
Cosmetic changes and missing data when CONFIG_PM=y. Otherwise lgtm.

File: arch/arm/src/stm32h7/stm32_serial.c (hunk 4 above)
Comment (davids5) [reply arch/arm/src/stm32h7/stm32_serial.c:3626]: @raiden00pl - thank you for the thorough review! I will squash the formatting changes and fix the PM or removed it.

Comment (davids5) [reply arch/arm/src/stm32h7/stm32_serial.c:3870]: @raiden00pl - I pulled in the PM from the F7 which is closer to one that works and have marked it untested.

Comment (davids5) [reply arch/arm/src/stm32h7/stm32_serial.c:3870]: It is squashed ...
```
stm32h7:Serial Apply formatting suggestions from code review

Co-authored-by: Mateusz Szafoni <raiden00pl@gmail.com>

stm32h7: Serail Add Power Managment (Untested)
```

**Review by raiden00pl — APPROVED** [review]
lgtm

**Pull request merged**